pkg/scheduler/objects/allocation.go

```diff
@@ -565,3 +567,39 @@ func (a *Allocation) setUserQuotaCheckPassed() {
 		a.askEvents.SendRequestFitsInUserQuota(a.allocationKey, a.applicationID, a.allocatedResource)
 	}
 }
+
+func (a *Allocation) setPreemptionPreConditionsCheckPassed() {
+	a.Lock()
+	defer a.Unlock()
+	if a.preemptionPreConditionsCheckFailed {
+		a.preemptionPreConditionsCheckFailed = false
+		a.askEvents.SendPreemptionPreConditionsCheckPassed(a.allocationKey, a.applicationID, a.allocatedResource)
+	}
+}
+
+func (a *Allocation) setPreemptionPreConditionsCheckFailed() {
+	a.Lock()
+	defer a.Unlock()
+	if !a.preemptionPreConditionsCheckFailed {
+		a.preemptionPreConditionsCheckFailed = true
+		a.askEvents.SendPreemptionPreConditionsCheckFailed(a.allocationKey, a.applicationID, a.allocatedResource)
+	}
+}
+
+func (a *Allocation) setPreemptionQueueGuaranteesCheckPassed() {
+	a.Lock()
+	defer a.Unlock()
+	if a.preemptionQueueGuaranteesCheckFailed {
+		a.preemptionQueueGuaranteesCheckFailed = false
+		a.askEvents.SendPreemptionQueueGuaranteesCheckPassed(a.allocationKey, a.applicationID, a.allocatedResource)
+	}
+}
+
+func (a *Allocation) setPreemptionQueueGuaranteesCheckFailed() {
+	a.Lock()
+	defer a.Unlock()
+	if !a.preemptionQueueGuaranteesCheckFailed {
+		a.preemptionQueueGuaranteesCheckFailed = true
+		a.askEvents.SendPreemptionQueueGuaranteesCheckFailed(a.allocationKey, a.applicationID, a.allocatedResource)
+	}
+}
```

Comment:
These are all `REQUEST` events which means they will be sent to the respective pods. 

I think the only message which might be useful is an indication that a particular pod successfully triggered preemption. Or the other way around: we send a message to the terminated pod that it's been preempted to make room for pod X.

Everything else (guarantees passed/failed, precondition passed/failed) is just too much noise which should not be sent to K8s as a pod event.

Reply:
I agree we should not send events except in these cases:

Request pod: "Preempting pod {namespace}/{pod} to make room"  (there could be multiple of these if we preempt multiple pods)

Victim pods: "Preempting pod to make room for {namespace}/{pod}" (only one of these per victim)

This allows users to see that their pod triggered preemption, or that their pod was preempted and by whom.

Reply:
As of now, there is only one log line "Reserving node for ask after preemption" being logged at the end of the preemption process. Other than this, we don't have any info. 

We need below info to debug the preemption related issues:

1. How do we know whether preemption has been even attempted or not?
2. If attempted, Is there any problem in passing through important steps?
3. If attempted and successful case, how do we know this case? Answer: Above log line helps

I went through all these questions while trying to find the reason for test failures logged in https://issues.apache.org/jira/browse/YUNIKORN-2808.

> I think the only message which might be useful is an indication that a particular pod successfully triggered preemption.

Yes. It definitely helps. We need something like this. I had also similar thought, but doing this way would always sends an event. Hence, I followed the other usual model of 1. Sending only one event and that too only when failed. Does not repeat. 2. Sending only one event when succeeded (Only If 1st event happens). Does not repeat.

> Request pod: "Preempting pod {namespace}/{pod} to make room" (there could be multiple of these if we preempt multiple pods)
> 
> Victim pods: "Preempting pod to make room for {namespace}/{pod}" (only one of these per victim)

Yes, it helps but only in successful case and that too at the end.

Reply:
We *must not* log for every attempt as this happens far too frequently. It will overwhelm the system. Currently in the scheduler we only log *when action is taken or the state of the system changes*.

Reply:
Pod events definitely should only be sent once, and only when action is taken.

Reply:
I think an acceptable approach is recording stuff like the allocation log for each request. That looks reasonable to me. We can have a type like `AllocationLog` and we define fixed causes about why the preemption failed:
* "Preconditions checks failed" (`preemptor.CheckPreconditions()` false)
* "No guarantee to free up resources" (`p.checkPreemptionQueueGuarantees()` false)
* "Preemption shortfall" (`p.ask.GetAllocatedResource().StrictlyGreaterThanOnlyExisting()` false)

Every time something fails, you increase a counter for it. You also count the number of total attempts.

So I'm thinking sth like:
```
// Mutable fields which need protection
allocated           bool
allocLog            map[string]*AllocationLogEntry
preemptionLog map[string]*AllocationLogEntry  // new field
preemptionAttempts int64  // new field
preemptionTriggered bool
preemptCheckTime    time.Time
...

var (
  ErrPreemptionPrecondFailed = errors.New("Preconditions checks failed")
  ErrPreemptionNoGuarantee = errors.New("No guarantee to free up resources")
  ErrPreemptionShortfall = errors.New("Preemption shortfall")
)

```
Then in `application.go`:
```
				if fullIterator != nil {
                                        request.IncPreemptionAttempts()
					if result, err := sa.tryPreemption(headRoom, preemptionDelay, request, fullIterator, false); err != nil {
						// preemption occurred, and possibly reservation
						return result
					}
					request.LogPreemptionFailure(err) 
				}
			}
```

Obviously you need to make sure that problems propagate back so you can log it. Then this whole thing can be exposed in the REST API just like the allocation log.

It's obviously not as detailed as a separate log entry printed to the stdout, but I see this as a good compromise.

Reply:
Why not just reuse `AllocationLog`? It already contains events pertaining to things like predicate failures; preemption attempts can be tracked there easily as well.

Reply:
Yeah, that's fine too.

Reply:
Have used allocationLog earlier too in few places in addition with events & logging. Now using only allocationLog, removed events and logs. Still seeing a gap to add debug logs in few places especially after guarantee checks passes. Anyways, we can  address this later depends on the need.